# Random Forests

In [29]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split as tts
from sklearn.model_selection import cross_val_score as cvs
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier

In [3]:
df = pd.read_csv('pima-indians-diabetes.csv')
df.columns = ['pregnancies', 'glucose', 'blood_pressure', 'skin_thickness', 'insulin', 'bmi', 'diabetes_pedigree_function', 'age', 'outcome']
df.head()

pregnancies  glucose  blood_pressure  skin_thickness  insulin   bmi  \
0            1       85              66              29        0  26.6   
1            8      183              64               0        0  23.3   
2            1       89              66              23       94  28.1   
3            0      137              40              35      168  43.1   
4            5      116              74               0        0  25.6   

   diabetes_pedigree_function  age  outcome  
0                       0.351   31        0  
1                       0.672   32        1  
2                       0.167   21        0  
3                       2.288   33        1  
4                       0.201   30        0

## Identify null values

In [4]:
df.isnull().sum()

pregnancies                   0
glucose                       0
blood_pressure                0
skin_thickness                0
insulin                       0
bmi                           0
diabetes_pedigree_function    0
age                           0
outcome                       0
dtype: int64

## Describe

In [5]:
df.describe()

pregnancies     glucose  blood_pressure  skin_thickness     insulin  \
count   767.000000  767.000000      767.000000      767.000000  767.000000   
mean      3.842243  120.859192       69.101695       20.517601   79.903520   
std       3.370877   31.978468       19.368155       15.954059  115.283105   
min       0.000000    0.000000        0.000000        0.000000    0.000000   
25%       1.000000   99.000000       62.000000        0.000000    0.000000   
50%       3.000000  117.000000       72.000000       23.000000   32.000000   
75%       6.000000  140.000000       80.000000       32.000000  127.500000   
max      17.000000  199.000000      122.000000       99.000000  846.000000   

              bmi  diabetes_pedigree_function         age     outcome  
count  767.000000                  767.000000  767.000000  767.000000  
mean    31.990482                    0.471674   33.219035    0.348110  
std      7.889091                    0.331497   11.752296    0.476682  
min      0.000000                    0.078000   21.000000    0.000000  
25%     27.300000                    0.243500   24.000000    0.000000  
50%     32.000000                    0.371000   29.000000    0.000000  
75%     36.600000                    0.625000   41.000000    1.000000  
max     67.100000                    2.420000   81.000000    1.000000

## Assess balance

In [6]:
df.outcome.value_counts()

outcome
0    500
1    267
Name: count, dtype: int64

500 samples without diabetes, 267 with diabetes. The data set is not too unbalanced, with a non-diabetic to diabetic ratio of about 2:1

## Assign X and y

In [7]:
X = df.drop('outcome', axis=1)
y = df.outcome

## Scale data

In [10]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled[:3]

array([[-0.84372629, -1.12208597, -0.16024856,  0.53202348, -0.69355921,
        -0.68372895, -0.36426474, -0.18894038],
       [ 1.23423997,  1.94447577, -0.26357823, -1.28688187, -0.69355921,
        -1.10230105,  0.60470064, -0.1037951 ],
       [-0.84372629, -0.99692019, -0.16024856,  0.15569823,  0.12235685,
        -0.49346891, -0.91968415, -1.0403932 ]])

## Train-Test Split

In [16]:
# Stratify argument helps to balance the data
X_train, X_test, y_train, y_test = tts(X_scaled, y, stratify=y, random_state=10)

In [18]:
X_train.shape

(575, 8)

In [19]:
X_test.shape

(192, 8)

In [20]:
y_train.value_counts()

outcome
0    375
1    200
Name: count, dtype: int64

## Create a decision tree

In [31]:
dt_classifier = DecisionTreeClassifier()
scores = cvs(dt_classifier, X, y, cv=5)
scores

array([0.71428571, 0.67532468, 0.69281046, 0.77124183, 0.76470588])

In [32]:
scores.mean()

np.float64(0.723673711909006)

In [ ]:

BaggingClassifier(
    base__estimator=DecisionTreeClassifier(),
    n_estimators=100,
    max_sampels=0.8,
    oob_score=True
    
)